[c10d] Increment sequence numbers on collectives.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55718 [c10d] Increment sequence numbers on collectives.**

Increments sequence numbers when ProcessGroupGloo::enqueue or
ProcessGroupNCCL::collective is run, which is a common call all collectives
make. The next step will be to log these along with other collective info in
debug mode as well as integrating them with the process group wrapper.

Differential Revision: [D27690690](https://our.internmc.facebook.com/intern/diff/D27690690/)

rohan-varma committed:
[c10d] Increment sequence numbers on collectives.

Increments sequence numbers when ProcessGroupGloo::enqueue or
ProcessGroupNCCL::collective is run, which is a common call all collectives
make. The next step will be to log these along with other collective info in
debug mode as well as integrating them with the process group wrapper.

Differential Revision: [D27690690](https://our.internmc.facebook.com/intern/diff/D27690690/)

[ghstack-poisoned]

SciPioneer commented on test/distributed/test_c10d.py:
```diff
@@ -4846,6 +4851,126 @@ def test_broadcast_coalesced_gloo_cpu(self):
+
+        if dist.get_world_size(process_group) > 2:
+            # Test when certain ranks don't call collectives
+            if dist.get_rank(process_group) not in [0, 1]:
```
It may be better to also add a non-contiguous number like 3 in the list of `[0, 1]`.

SciPioneer commented on test/distributed/test_c10d.py:
```diff
@@ -4846,6 +4851,126 @@ def test_broadcast_coalesced_gloo_cpu(self):
+
+    @skip_if_lt_x_gpu(4)
+    @requires_nccl()
+    def test_sequence_num_incremented_nccl_subgroup(self):
```
Nit: Create a helper function to avoid the boilerplate code here, which is same as `test_sequence_num_incremented_gloo_subgroup`? It seems that only one arg differs in these two tests.

wayi1 reviewed (approved):
Thanks for improving the debuggability!

rohan-varma committed:
Update on "[c10d] Increment sequence numbers on collectives."


Increments sequence numbers when ProcessGroupGloo::enqueue or
ProcessGroupNCCL::collective is run, which is a common call all collectives
make. The next step will be to log these along with other collective info in
debug mode as well as integrating them with the process group wrapper.

Differential Revision: [D27690690](https://our.internmc.facebook.com/intern/diff/D27690690/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Increment sequence numbers on collectives."


Increments sequence numbers when ProcessGroupGloo::enqueue or
ProcessGroupNCCL::collective is run, which is a common call all collectives
make. The next step will be to log these along with other collective info in
debug mode as well as integrating them with the process group wrapper.

Differential Revision: [D27690690](https://our.internmc.facebook.com/intern/diff/D27690690/)

[ghstack-poisoned]

zhaojuanmao commented on test/distributed/test_c10d.py:
```diff
@@ -4905,6 +4909,126 @@ def test_broadcast_coalesced_gloo_cpu(self):
+        )
+        assert dist.get_world_size(process_group) == dist.get_world_size(verify_pg)
+
+        initial_num = (
+            self._verify_sequence_number_across_pg(
+                pg=process_group, verify_pg=verify_pg
+            )
+            if not c10d.distributed_c10d._rank_not_in_group(process_group)
+            else -1
+        )
```
nit: validate initial_num is the same across the ranks

rohan-varma replied:
The call to `_verify_sequence_number_across_pg` handles this

rohan-varma committed:
Update on "[c10d] Increment sequence numbers on collectives."


Increments sequence numbers when ProcessGroupGloo::enqueue or
ProcessGroupNCCL::collective is run, which is a common call all collectives
make. The next step will be to log these along with other collective info in
debug mode as well as integrating them with the process group wrapper.

Differential Revision: [D27690690](https://our.internmc.facebook.com/intern/diff/D27690690/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Increment sequence numbers on collectives."


Increments sequence numbers when ProcessGroupGloo::enqueue or
ProcessGroupNCCL::collective is run, which is a common call all collectives
make. The next step will be to log these along with other collective info in
debug mode as well as integrating them with the process group wrapper.

Differential Revision: [D27690690](https://our.internmc.facebook.com/intern/diff/D27690690/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Increment sequence numbers on collectives."


Increments sequence numbers when ProcessGroupGloo::enqueue or
ProcessGroupNCCL::collective is run, which is a common call all collectives
make. The next step will be to log these along with other collective info in
debug mode as well as integrating them with the process group wrapper.

Differential Revision: [D27690690](https://our.internmc.facebook.com/intern/diff/D27690690/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Increment sequence numbers on collectives."


Increments sequence numbers when ProcessGroupGloo::enqueue or
ProcessGroupNCCL::collective is run, which is a common call all collectives
make. The next step will be to log these along with other collective info in
debug mode as well as integrating them with the process group wrapper.

Differential Revision: [D27690690](https://our.internmc.facebook.com/intern/diff/D27690690/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Increment sequence numbers on collectives."


Increments sequence numbers when ProcessGroupGloo::enqueue or
ProcessGroupNCCL::collective is run, which is a common call all collectives
make. The next step will be to log these along with other collective info in
debug mode as well as integrating them with the process group wrapper.

Differential Revision: [D27690690](https://our.internmc.facebook.com/intern/diff/D27690690/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Increment sequence numbers on collectives."


Increments sequence numbers when ProcessGroupGloo::enqueue or
ProcessGroupNCCL::collective is run, which is a common call all collectives
make. The next step will be to log these along with other collective info in
debug mode as well as integrating them with the process group wrapper.

Differential Revision: [D27690690](https://our.internmc.facebook.com/intern/diff/D27690690/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Increment sequence numbers on collectives."


Increments sequence numbers when ProcessGroupGloo::enqueue or
ProcessGroupNCCL::collective is run, which is a common call all collectives
make. The next step will be to log these along with other collective info in
debug mode as well as integrating them with the process group wrapper.

Differential Revision: [D27690690](https://our.internmc.facebook.com/intern/diff/D27690690/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] Increment sequence numbers on collectives."


Increments sequence numbers when ProcessGroupGloo::enqueue or
ProcessGroupNCCL::collective is run, which is a common call all collectives
make. The next step will be to log these along with other collective info in
debug mode as well as integrating them with the process group wrapper.

Differential Revision: [D27690690](https://our.internmc.facebook.com/intern/diff/D27690690/)

[ghstack-poisoned]

rohan-varma commented:
ROCM succeeded - https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.1-py3.6-build/5174/console
